pkg/runtime/wfengine/activity.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	// Try to load activity state. If we find any, that means the activity invocation is a duplicate.
+	if state, err := a.loadActivityState(ctx, actorID, ar.Generation); err != nil {
+		return nil, err
+	} else if state.Generation > 0 {
+		return nil, ErrDuplicateInvocation
+	}
+
+	// Save the request details to the state store in case we need it after recovering from a failure.
+	state := activityState{
+		Generation:   ar.Generation,
+		EventPayload: ar.HistoryEvent,
+	}
+	if err := a.saveActivityState(ctx, actorID, state); err != nil {
+		return nil, err
+	}
+
+	// The actual execution is triggered by a reminder
+	err := a.createReliableReminder(ctx, actorID, ar.Generation)
+	return nil, err
+}
+
+// InvokeReminder implements actors.InternalActor and executes the activity logic.
+func (a *activityActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, data []byte, dueTime string, period string) error {
+	wfLogger.Debugf("invoking reminder '%s' on activity actor '%s'", reminderName, actorID)
+
+	var generation uint64
+	if err := actors.DecodeInternalActorReminderData(data, &generation); err != nil {
+		// Likely the result of an incompatible activity reminder format change. This is non-recoverable.
+		return err
+	}
+	state, _ := a.loadActivityState(ctx, actorID, generation)
+	// TODO: On error, reply with a failure - this requires support from durabletask-go to produce TaskFailure results
+
+	timeoutCtx, cancelTimeout := context.WithTimeout(ctx, a.defaultTimeout)
+	defer cancelTimeout()
+
+	if err := a.executeActivity(timeoutCtx, actorID, reminderName, state.EventPayload); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			wfLogger.Warnf("%s: execution of '%s' timed-out and will be retried later", actorID, reminderName)
+
+			// Returning nil signals that we want the execution to be retried in the next period interval
+			return nil
+		} else if _, ok := err.(recoverableError); ok {
+			wfLogger.Warnf("%s: execution failed with a recoverable error and will be retried later: %v", actorID, err)
+
+			// Returning nil signals that we want the execution to be retried in the next period interval
+			return nil
+		} else {
+			wfLogger.Errorf("%s: execution failed with a non-recoverable error: %v", actorID, err)
+			// TODO: Reply with a failure - this requires support from durabletask-go to produce TaskFailure results
+		}
+	}
+
+	// TODO: Purge actor state based on some data retention policy
+
+	// We delete the reminder on success and on non-recoverable errors.
+	return actors.ErrReminderCanceled
+}
+
+func (a *activityActor) executeActivity(ctx context.Context, actorID string, name string, eventPayload []byte) error {
+	taskEvent, err := backend.UnmarshalHistoryEvent(eventPayload)
+	if err != nil {
+		return err
+	}
+
+	endIndex := strings.LastIndex(actorID, "#")
+	if endIndex < 0 {
+		return fmt.Errorf("invalid activity actor ID: %s", actorID)
+	}
+	workflowID := actorID[0:endIndex]
```

Comment:
qq: Do you think worth to have an ActorID type?

`type ActorID = string` 

so you can have custom operations like


```golang
func (id ActorID) WorkflowID() (string, error) {
...
}
```

Reply:
plus: A custom constructor that can construct/deconstruct using the desired format

Reply:
I think it's worth having an `ActorID` type if we want to enforce specific constraints or operations on those IDs. I don't really consider myself an expert on actors, though, so I might leave something like that up to an Actors SME. It may be best to consider this outside the scope of this PR, however.

Reply:
Just to clarify, I guess the correct type would be `type internalActorID` as you've introduced the logic behind using a specific actor ID for workflows and activities.

The idea is to have a single place where the internal actor id are built to make clear that this is a special ID that contains the workflow id on it

Reply:
We don't build internal actor IDs anywhere in the Dapr codebase. They are all generated externally, in this case either by consumers of the workflow API/SDK or via the `durabletask-go` dependency in cases where the workflow orchestration IDs are auto-generated UUIDs. I don't think having an `internalActorID` type will serve much purpose, at least not right now.

Reply:
Even worse, in my opinion. We make an assumption based on a string derived from a library/sdk implementation. I believe that the ideal is to have this resolved on the sdk side, since it is responsible for generating the ID.

My general comment here is that the information on how the ID is formatted (which character should be used, what are the first part of the id and the second) is not unified in one place, but rather spread out between the creator and the consumer. Due to the lack of a well-defined type, this is passed as simply a string and its deconstruction is not clear to anyone reading the code. In case the sdk/library decides to have more information on that ID, it would be very difficult to modify it without a lot of effort.

For now, this comment doesn't mean much as this is the only place where the ID is being deconstructed, but if this starts to increase it could be a problem. In that case, I don't know if you and I would be here to warn about it.

Reply:
just to make sure we are on the same page, I'm talking about [this](https://github.com/dapr/dapr/pull/5784/files/f7accc5308949f4c6d152e439e5dfb7495b47932#diff-092e8c8daaca7a6f737bf96a9b901f50a8f0f4a4fa2183ef14b8b4cfd726ed5aR145)

Reply:
Ah, I think I was misunderstanding your root concern. You're specifically talking about the convention for activity actor IDs, which is {workflowID}#{taskID}. This ID is generated from a function [here](https://github.com/dapr/dapr/pull/5784/files/f7accc5308949f4c6d152e439e5dfb7495b47932#diff-9997677e8aa2146ca3f09e604ca0f932a17a85fddc758fcc785c7ebb9afbc08aR457-R460). Would you be okay with me encapsulating this deconstruction logic in a function as well?

Reply:
Sorry, my comment was wrongly pointed the line where the actor id is used. But yes, I will be fine with that, go ahead and create the function